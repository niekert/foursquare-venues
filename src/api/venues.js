import { apiRequest } from './request';

export const fetchVenues = async ({
  latLong,
  near,
  radius = 250,
  venuePhotos = 1,
}) =>
  apiRequest('/venues/explore', {
    ll: latLong,
    near,
    venuePhotos,
    radius,
  }).then(data => data.response);
