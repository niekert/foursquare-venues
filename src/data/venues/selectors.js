export function getVenuesList(venuesState) {
  // For now, we always select the venues in the first group. This could be changed
  // If we extend the app's capabilities
  return venuesState.venues && venuesState.venues.groups
    ? venuesState.venues.groups[0].items
    : [];
}

/**
 * Get the recommended map bounds in a literal that the google maps API supports
 * https://developers.google.com/maps/documentation/javascript/3.exp/reference#LatLngBoundsLiteral
 * @param {*} venuesState
 */
export function getMapBoundsLiteral(venuesState) {
  if (!venuesState.venues) {
    return null;
  }

  const { ne, sw } = venuesState.venues.suggestedBounds;
  return {
    east: ne.lng,
    north: ne.lat,
    south: sw.lat,
    west: sw.lng,
  };
}
