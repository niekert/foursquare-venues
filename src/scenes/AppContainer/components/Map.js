import React from 'react';
import styled from 'styled-components';
import { FullScreenMap, Loading } from 'shared/components';

function Map() {
  return (
    <FullScreenMap
      loadingElement={<div />}
      defaultZoom={8}
      defaultCenter={{ lat: -34.397, lng: 150.644 }}
    />
  );
}

export default Map;
