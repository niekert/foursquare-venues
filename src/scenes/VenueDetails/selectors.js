import { get, take } from 'lodash';
/**
 * Get the urls of the photo of a venue in a given size
 * @param {*} venue the venue to get photos of
 * @param {*} amount the amount of photos
 * @param {*} size the size of each foto
 */
export function getPhotoUrls(venue, amount = 6, size = '100x100') {
  const photos = get(venue, 'photos.groups[0].items', []);
  const sliced = take(photos, amount);

  return sliced.map(photo => `${photo.prefix}${size}${photo.suffix}`);
}
