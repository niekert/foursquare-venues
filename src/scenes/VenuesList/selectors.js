import { get } from 'lodash-es';

export const getThumbnailUrl = (venue, size = '50x50') => {
  const thumbnailDetails = get(venue, 'photos.groups[0].items[0]');

  return thumbnailDetails
    ? `${thumbnailDetails.prefix}${size}${thumbnailDetails.suffix}`
    : null;
};

/**
 * Get the first category a venue is placed in
 * @param {*} venue The venues category
 */
export const getCategory = venue =>
  venue.categories.length > 0 ? venue.categories[0] : null;
