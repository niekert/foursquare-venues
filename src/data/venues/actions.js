import { fetchVenues } from 'api/venues';
import 'babel-polyfill';

export const FETCH_VENUES_BY_GEO = 'FETCH_VENUES_BY_GEO';
export const SELECT_VENUE = 'SELECT_VENUE';
export const CLOSE_SELECTED_VENUE = 'CLOSE_SELECTED_VENUE';
export const FETCH_VENUES_BY_SEARCH = 'FETCH_VENUES_BY_SEARCH';
export const FETCH_VENUES_SUCCESS = 'FETCH_VENUES_SUCCESS';
export const FETCH_VENUES_ERROR = 'FETCH_VENUES_ERROR';

export function fetchVenuesByGeoLocation(position) {
  const { coords: { latitude, longitude } } = position;
  const latLng = {
    lat: latitude,
    lng: longitude,
  };
  return async dispatch => {
    dispatch({
      type: FETCH_VENUES_BY_GEO,
      payload: {
        latLng,
      },
    });

    const venues = await fetchVenues({
      latLong: [latitude, longitude].join(','),
    });

    dispatch({
      type: FETCH_VENUES_SUCCESS,
      payload: {
        venues,
      },
    });
  };
}

export function fetchVenuesBySearch(searchQuery) {
  return async dispatch => {
    dispatch({
      type: FETCH_VENUES_BY_SEARCH,
      payload: {
        query: searchQuery,
      },
    });

    const venues = await fetchVenues({
      near: searchQuery,
    });

    dispatch({
      type: FETCH_VENUES_SUCCESS,
      payload: {
        venues,
      },
    });
  };
}

export function selectVenue(venueId) {
  return {
    type: SELECT_VENUE,
    payload: {
      venueId,
    },
  };
}

export function closeSelectedVenue() {
  return {
    type: CLOSE_SELECTED_VENUE,
  };
}
