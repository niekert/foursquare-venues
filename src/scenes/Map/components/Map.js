import React, { Component } from 'react';
import { arrayOf, func, string, shape, number } from 'prop-types';
import { latLong, venue as venuePropType } from 'customPropTypes';
import { getVenueId } from 'data/venues/selectors';
import { MyLocationMarker } from 'style/MapMarkers';
import {
  DEFAULT_MAP_LAT_LNG,
  DEFAULT_MAP_ZOOM,
  SELECTED_MAP_ZOOM,
} from 'app-constants';
import { FullScreenMap, Loading } from 'shared/components';
import VenueMarker from './VenueMarker';

class Map extends Component {
  static propTypes = {
    selectVenue: func.isRequired,
    mapBounds: shape({
      north: number.isRequired,
      east: number.isRequired,
      west: number.isRequired,
      south: number.isRequired,
    }),
    closeSelectedVenue: func.isRequired,
    selectedVenueId: string,
    currentLocationLatLng: latLong,
    venues: arrayOf(venuePropType),
  };

  static defaultProps = {
    currentLocationLatLng: null,
    selectedVenueId: null,
    mapBounds: null,
    venues: [],
  };

  componentDidUpdate(prevProps) {
    if (
      this.mapElement &&
      this.props.mapBounds &&
      // TODO: Don't use stringify but use a proper shallow compare
      JSON.stringify(prevProps.mapBounds) !==
        JSON.stringify(this.props.mapBounds)
    ) {
      this.mapElement.fitBounds(this.props.mapBounds);
    }
  }

  render() {
    const {
      currentLocationLatLng,
      selectedVenueId,
      closeSelectedVenue,
      selectVenue,
      venues,
    } = this.props;

    return (
      <FullScreenMap
        onMapMounted={map => {
          this.mapElement = map;
        }}
        loadingElement={<Loading />}
        defaultZoom={DEFAULT_MAP_ZOOM}
        zoom={selectedVenueId ? SELECTED_MAP_ZOOM : DEFAULT_MAP_ZOOM}
        defaultCenter={DEFAULT_MAP_LAT_LNG}
      >
        {currentLocationLatLng && (
          <MyLocationMarker position={currentLocationLatLng} />
        )}
        {venues.map(venue => {
          const venueId = getVenueId(venue);
          return (
            <VenueMarker
              key={`venue-marker-${venueId}`}
              isDetailsOpen={selectedVenueId === venueId}
              closeDetails={closeSelectedVenue}
              venue={venue}
              onClick={() => selectVenue(venueId)}
            />
          );
        })}
      </FullScreenMap>
    );
  }
}

export default Map;
