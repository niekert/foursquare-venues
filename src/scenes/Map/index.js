import { connect } from 'react-redux';
import { selectVenue, closeSelectedVenue } from 'data/venues/actions';
import { getVenuesList, getMapBoundsLiteral } from 'data/venues/selectors';
import Map from './components/Map';

function mapStateToProps(state) {
  const { venues } = state.data;

  return {
    isLoading: venues.isLoading,
    currentLocationLatLng: venues.currentGeoLatLng,
    selectedVenueId: venues.selectedVenueId,
    mapBounds: getMapBoundsLiteral(venues),
    venues: getVenuesList(venues),
  };
}

export default connect(mapStateToProps, { selectVenue, closeSelectedVenue })(
  Map,
);
