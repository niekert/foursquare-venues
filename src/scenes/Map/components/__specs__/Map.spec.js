import React from 'react';
import * as venuesSelectors from 'data/venues/selectors';
import { shallow } from 'enzyme';
import Map from '../Map';

describe('<Map />', () => {
  let sharedProp;
  beforeEach(() => {
    venuesSelectors.getVenueId = () => 'mock-venue-id';

    sharedProp = {
      currentLocationLatLng: {
        lat: 52,
        lng: 53,
      },
      selectedVenueId: 'mock_venue',
      venues: ['fake-venue-1', 'fake-venue-2', 'fake-venue-3', 'mock-venue-id'],
    };
  });

  test('renders correctly', () => {
    const wrapper = shallow(<Map {...sharedProp} />);
    expect(wrapper).toMatchSnapshot();
  });
});
