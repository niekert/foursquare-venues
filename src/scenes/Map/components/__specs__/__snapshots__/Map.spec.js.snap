// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Map /> renders correctly 1`] = `
<FullScreenMap
  defaultCenter={
    Object {
      "lat": 52.3702,
      "lng": 4.8952,
    }
  }
  defaultZoom={12}
  loadingElement={<Loading />}
  onMapMounted={[Function]}
  zoom={16}
>
  <MapMarkers__MyLocationMarker
    position={
      Object {
        "lat": 52,
        "lng": 53,
      }
    }
  />
  <VenueMarker
    isDetailsOpen={false}
    key="venue-marker-mock-venue-id"
    onClick={[Function]}
    venue="fake-venue-1"
  />
  <VenueMarker
    isDetailsOpen={false}
    key="venue-marker-mock-venue-id"
    onClick={[Function]}
    venue="fake-venue-2"
  />
  <VenueMarker
    isDetailsOpen={false}
    key="venue-marker-mock-venue-id"
    onClick={[Function]}
    venue="fake-venue-3"
  />
  <VenueMarker
    isDetailsOpen={false}
    key="venue-marker-mock-venue-id"
    onClick={[Function]}
    venue="mock-venue-id"
  />
</FullScreenMap>
`;
